bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java

```diff
@@ -0,0 +1,278 @@
+/**
+ * Copyright (c) 2014-2015 openHAB UG (haftungsbeschraenkt) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.core.common;
+
+import java.text.ParseException;
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.Collections;
+import java.util.Date;
+import java.util.StringTokenizer;
+import java.util.TimeZone;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * <code>AbstractExpression</code> is an abstract implementation of {@link Expression} that provides common
+ * functionality to other concrete implementations of <code>Expression</code>
+ *
+ * @author Karel Goderis - Initial Contribution
+ *
+ */
+public abstract class AbstractExpression<E extends AbstractExpressionPart> implements Expression {
+
+    private final Logger logger = LoggerFactory.getLogger(this.getClass().getName());
+
+    private int minimumCandidates = 10;
+    private int maximumCandidates = 100;
+
+    private String expression;
+    private String delimiters;
+    private ArrayList<E> expressionParts = new ArrayList<E>();
+
+    private boolean continueSearch;
+    private ArrayList<Date> candidates = new ArrayList<Date>();
+    private Date startDate = null;
+    private TimeZone timeZone = null;
+
+    /**
+     * Build an {@link Expression}
+     *
+     * @param expression - the expression
+     * @param delimiters - delimiters to consider when splitting the expression into expression parts
+     * @param startDate - the start date of the expression
+     * @param timeZone - the time zone of the expression
+     * @param minimumCandidates - the minimum number of candidate dates to calculate
+     * @throws ParseException - when the expression cannot be parsed correctly
+     */
+    public AbstractExpression(String expression, String delimiters, Date startDate, TimeZone timeZone,
+            int minimumCandidates) throws ParseException {
+        super();
+
+        if (expression == null) {
+            throw new IllegalArgumentException("The expression cannot be null");
+        }
+
+        this.expression = expression;
+        this.delimiters = delimiters;
+        this.startDate = startDate;
+        this.timeZone = timeZone;
+        this.minimumCandidates = minimumCandidates;
+
+        setStartDate(startDate);
+        setTimeZone(timeZone);
+        parseExpression(expression);
+    }
+
+    @Override
+    public final Date getStartDate() {
+        if (startDate == null) {
```

Comment:
All fine, but I would assume that the getStartDate is called more often (e.g. more often than setStartDate).
If we ensure that startDate is never null (in the constructor and setStartDate function) we do not need this check every time getStartDate is called.

Or do you like that the time of the first getStartDate call is used?